pkg/cmd/cli/cmd/edit.go

```diff
@@ -0,0 +1,333 @@
+	cmd.Flags().String("output-version", "", "Output the formatted object with the given version (default api-version).")
+	cmd.Flags().VarP(&filenames, "filename", "f", "Filename, directory, or URL to file to use to edit the resource.")
+	return cmd
+}
+
+func RunEdit(fullName string, f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args []string, filenames util.StringList) error {
+	var p kubectl.ResourcePrinter
+	switch format := cmdutil.GetFlagString(cmd, "output"); format {
+	case "json":
+		p = &kubectl.JSONPrinter{}
+	case "yaml":
+		p = &kubectl.YAMLPrinter{}
+	default:
+		return cmdutil.UsageError(cmd, "The flag 'output' must be one of yaml|json")
+	}
+
+	cmdNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	mapper, typer := f.Object()
+	rmap := &resource.Mapper{
+		ObjectTyper:  typer,
+		RESTMapper:   mapper,
+		ClientMapper: f.ClientMapperForCommand(),
+	}
+
+	b := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(filenames...).
+		//SelectorParam(selector).
+		ResourceTypeOrNameArgs(true, args...).
+		Latest()
+	if err != nil {
+		return err
+	}
+
+	clientConfig, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	r := b.Flatten().Do()
+	obj, err := r.Object()
+	if err != nil {
+		return err
+	}
+
+	results := editResults{}
+	for {
+		// try conversion to all the possible versions
+		// TODO: simplify by adding a ResourceBuilder mode
+		defaultVersion := clientConfig.Version
+		// the outermost object will be converted to the output-version
+		version := cmdutil.OutputVersion(cmd, defaultVersion)
+		versions := []string{version, latest.Version}
+		infos, _ := r.Infos()
+		for _, info := range infos {
+			versions = append(versions, info.Mapping.APIVersion)
+		}
+
+		// TODO: add a new ResourceBuilder mode for Object() that attempts to ensure the objects
+		//   are in the appropriate version if one exists (and if not, use the best effort).
+		// TODO: ensure api-version is set with the default preferred api version by the client
+		//   builder on initialization
+		// TODO: add an annotating YAML printer that can print inline comments on each field,
+		//   including descriptions or validation errors
+		printer := kubectl.NewVersionedPrinter(p, kapi.Scheme, versions...)
+
+		// generate the file to edit
+		buf := &bytes.Buffer{}
+		if err := results.header.WriteTo(buf); err != nil {
+			return err
+		}
+		if err := printer.PrintObj(obj, buf); err != nil {
+			return err
+		}
+		original := buf.Bytes()
+
+		// launch the editor
+		edit := editor.NewDefaultEditor()
+		edited, file, err := edit.LaunchTempFile("", "osc-edit-", buf)
+		if err != nil {
+			return err
+		}
+		glog.V(4).Infof("User edited:\n%s", string(edited))
+		lines, err := hasLines(bytes.NewBuffer(edited))
+		if err != nil {
+			return preservedFile(err, file, cmd.Out())
+		}
+		if bytes.Equal(original, edited) {
```

Comment:
If I edit improperly and end up back in this loop, then the second time I want to exit and rethink my life.  This deletes my file so I can't review.